# Selective RNA modification in situ by circular gRNAs in designer organelles

This repository contains the bioinformatics scripts used for the pseudouridine modification quantification based on direct RNA sequencing data of HEK293T+reporter transcriptomes in the following publication: 

Schartel, Jann, Wierczeiko, Butto, Helm, Gerber, Lemke - Selective RNA modification in situ by circular gRNAs in designer organelles

## Content
1. *Processing of raw data:*
In the DRS_processing.sh bash script, direct RNA sequencing (DRS) data is basecalled, mapped onto the reporter sequences and filtered for positions of interest.
2. *Accuracy and Pseudouridine modification quantification:*
The jupyter notebook Direct_RNA_seq_pattern_processing.ipynb contains the bascalling accuracy calculation of all samples and the subsequent pseudouridine quantification on specific positions by measuring the U>C mismatch frequency generated by the basecalling error and the dorado-based pseudouridine calling freqeuncy for remaining reads.
3. *Visualization*
We plotted the 10mer content around the modified positions in R.

For more information, please read the publication.

## References
[Schartel, Jann, Wierczeiko, Butto, Helm, Gerber, Lemke - Selective RNA modification in situ by circular gRNAs in designer organelles]
